File: Templates/Project Templates/macOS/Application Extension/Photo Project Extension.xctemplate/ProjectTypeDescriptionDataSource.m
//
//  ___FILENAME___
//  ___PACKAGENAME___
//
//  Created by ___FULLUSERNAME___ on ___DATE___.
//  ___COPYRIGHT___
//

#import "ProjectTypeDescriptionDataSource.h"

@implementation ProjectTypeDescriptionDataSource

+ (NSArray<PHProjectTypeDescription *> *)rootLevelProjectTypes {
    NSArray<PHProjectTypeDescription *> *projectTypes = @[];
    // TODO: Fill the array with PHProjectTypeDescription instances representing you project types for the root level.
    return projectTypes;
}

#pragma mark - PHProjectTypeDescriptionDataSource

- (nonnull NSArray<PHProjectTypeDescription *> *)subtypesForProjectType:(nonnull PHProjectType)projectType {
    if ([projectType isEqualToString:PHProjectTypeUndefined]) {
        return [self.class rootLevelProjectTypes];
    } else {
        NSArray<PHProjectTypeDescription *> *projectTypes = @[];
        // TODO: Fill the array with PHProjectTypeDescription instances representing you project types for the given level.
        return projectTypes;
    }
}

- (nullable PHProjectTypeDescription *)typeDescriptionForProjectType:(nonnull PHProjectType)projectType {
    // TODO: return the requested project type which was previously invalidated
    return nil;
}

- (nullable NSAttributedString *)footerTextForSubtypesOfProjectType:(nonnull PHProjectType)projectType {
    return nil;
}

@end

File: Templates/Project Templates/macOS/Application Extension/Photo Project Extension.xctemplate/ProjectTypeDescriptionDataSource.swift
//
//  ___FILENAME___
//  ___PACKAGENAME___
//
//  Created by ___FULLUSERNAME___ on ___DATE___.
//  ___COPYRIGHT___
//

import PhotosUI

class ProjectTypeDescriptionDataSource: NSObject, PHProjectTypeDescriptionDataSource {

    class var rootLevelProjectTypes: [PHProjectTypeDescription] {
        let projectTypes = [PHProjectTypeDescription]()
        // TODO: Fill the array with PHProjectTypeDescription instances representing you project types for the root level.
        return projectTypes
    }
    
    // MARK: - PHProjectTypeDescriptionDataSource
    func subtypes(for projectType: PHProjectType) -> [PHProjectTypeDescription] {
        if projectType == .undefined {
            return type(of: self).rootLevelProjectTypes
        } else {
            let projectTypes = [PHProjectTypeDescription]()
            // TODO: Fill the array with PHProjectTypeDescription instances representing you project types for the given level.
            return projectTypes
        }
    }
    
    func typeDescription(for projectType: PHProjectType) -> PHProjectTypeDescription? {
        // TODO: return the requested project type which was previously invalidated
        return nil
    }
    
    func footerText(forSubtypesOf projectType: PHProjectType) -> NSAttributedString? {
        return nil
    }
    
}
